File: cpp/graph/dijkstra-radix.cpp
#include<vector>
using namespace std;

#define B 33

template<typename T>
class RadixHeap {
  struct Val {
    int x; T key;

    bool operator<(const Val &other) {
      return x < other.x;
    }
  };
  vector<Val> data[B];
  int last = 0, siz = 0;

  inline int bsr(int x) {
    if(x == 0) return -1;
    return 31-__builtin_clz(x);
  }

public:
  inline void push(T key, int x) {
    //assert(last <= x);
    ++siz;
    data[bsr(x^last)+1].emplace_back(Val{x, key});
  }

  inline Val pop() {
    //assert(siz > 0);
    if(!data[0].size()) {
      int i = 1;
      while(!data[i].size()) ++i;
      Val &em = *min_element(data[i].begin(), data[i].end());
      for(Val &e : data[i]) {
        if(e.key == em.key) {
          --siz;
          continue;
        }
        data[bsr(e.x^em.x)+1].emplace_back(e);
      }
      data[i].clear();
      return em;
    }
    --siz;
    Val &em = data[0].back();
    data[0].pop_back();
    return em;
  }

  inline bool empty() {
    return siz == 0;
  }

  inline int size() {
    return siz;
  }
};

struct Edge {
  int to; int cost;
};

const int inf = 1e9+1;

#define N 100004
int n;
vector<Edge> g[N];
int dist[N];

void dijkstra(int s) {
  RadixHeap<int> que;
  que.push(s, 0);
  for(int i=0; i<n; ++i) dist[i] = inf;
  dist[s] = 0;

  while(!que.empty()) {
    auto p = que.pop();
    int co = p.x; int v = p.key;
    if(dist[v] < co) continue;
    for(auto &e : g[v]) {
      if(co + e.cost < dist[e.to]) {
        dist[e.to] = co + e.cost;
        que.push(e.to, dist[e.to]);
      }
    }
  }
}

File: python/graph/dijkstra-radix.py
class RadixHeap:
    __slots__ = ["data", "last", "siz", "used"]

    # (max_key - min_key) <= C
    def __init__(self, N, C):
        self.data = [[] for i in range(C.bit_length() + 1)]
        self.last = self.siz = 0
        self.used = [0]*N

    def push(self, x, key):
        #assert self.last <= x
        self.siz += 1
        self.data[(x ^ self.last).bit_length()].append((x, key))

    def pop(self):
        data = self.data
        used = self.used
        #assert self.siz > 0
        if not data[0]:
            i = 1
            while not data[i]:
                i += 1
            d = data[i]
            new_last, new_key = min(d)
            used[new_key] = 1
            for val in d:
                x, key = val
                if used[key]:
                    self.siz -= 1
                    continue
                data[(x ^ new_last).bit_length()].append(val)
            self.last = new_last
            data[i] = []
        else:
            new_last, new_key = data[0].pop()
            used[new_key] = 1
            self.siz -= 1
        return new_last, new_key

    def __len__(self):
        return self.siz

def dijkstra(N, G, s):
    que = RadixHeap(N, 10**9 + 1)

    dist = [10**18]*N
    dist[s] = 0
    que.push(0, s)
    while que:
        cost, v = que.pop()
        if dist[v] < cost:
            continue
        for w, c in G[v]:
            if cost + c < dist[w]:
                dist[w] = r = cost + c
                que.push(r, w)
    return dist
